src/FSharp.Control.TaskSeq/TaskSeq.fsi

```diff
@@ -154,414 +175,872 @@ module TaskSeq =
+    /// preventing such overlap to happen. Converting an array of tasks into a task sequence is no guarantee
+    /// that overlapping side effects are prevented. Safe for side-effect free tasks.
+    /// </summary>
+    ///
+    /// <param name="source">The input array-of-tasks.</param>
+    /// <returns>The resulting task sequence.</returns>
+    /// <exception cref="T:ArgumentNullException">Thrown when the input array is null.</exception>
     val ofTaskArray: source: #Task<'T> array -> taskSeq<'T>
 
-    /// Create a taskSeq of a seq of async.
+    /// <summary>
+    /// Views the given <see cref="seq" /> of <see cref="Async" />s as a task sequence, that is, as an
+    /// <see cref="IAsyncEnumerable&lt;'T>" />. A sequence of asyncs is not the same as a task sequence.
+    /// Each async computation in a sequence of asyncs can be run individually or in parallel, potentially
+    /// with overlapping side effects, while a task sequence forces awaiting between the items in the sequence,
+    /// preventing such overlap to happen.
+    /// </summary>
+    ///
+    /// <param name="source">The input sequence-of-asyncs.</param>
+    /// <returns>The resulting task sequence.</returns>
+    /// <exception cref="T:ArgumentNullException">Thrown when the input sequence is null.</exception>
     val ofAsyncSeq: source: seq<Async<'T>> -> taskSeq<'T>
 
-    /// Create a taskSeq of a list of async.
+    /// <summary>
+    /// Views the given <see cref="list" /> of <see cref="Async" />s as a task sequence, that is, as an
+    /// <see cref="IAsyncEnumerable&lt;'T>" />. A list of asyncs is not the same as a task sequence.
+    /// Each async computation in a list of asyncs can be run individually or in parallel, potentially
+    /// with overlapping side effects, while a task sequence forces awaiting between the items in the sequence,
+    /// preventing such overlap to happen.
+    /// </summary>
+    ///
+    /// <param name="source">The input list-of-asyncs.</param>
+    /// <returns>The resulting task sequence.</returns>
     val ofAsyncList: source: Async<'T> list -> taskSeq<'T>
 
-    /// Create a taskSeq of an array of async.
+    /// <summary>
+    /// Views the given <see cref="array" /> of <see cref="Async" />s as a task sequence, that is, as an
+    /// <see cref="IAsyncEnumerable&lt;'T>" />. An array of asyncs is not the same as a task sequence.
+    /// Each async computation in an array of asyncs can be run individually or in parallel, potentially
+    /// with overlapping side effects, while a task sequence forces awaiting between the items in the sequence,
+    /// preventing such overlap to happen.
+    /// </summary>
+    ///
+    /// <param name="source">The input array-of-asyncs.</param>
+    /// <returns>The resulting task sequence.</returns>
+    /// <exception cref="T:ArgumentNullException">Thrown when the input sequence is null.</exception>
     val ofAsyncArray: source: Async<'T> array -> taskSeq<'T>
```

Comment:
why is there an array, list and a seq variant? i.e. Async.Parallel takes a seq and then immediately calls toArray because it needs that internally, and life goes on
In most cases I'll pass an array comprehension as an argument, but it's nice to not have to do a ToArray if I happen to have a seq or list to hand
The more such overloads we have with either identical long xmldoc, or subtly varying ones, the more
- some type of people spend time navigating the intellisense or docs figuring out why there are N of them in order to make the correct/optimal choice
- other kinds of people will just arbitrarily pick one (e.g. if you have a pipeline that starts with a List, keep using List.map when the final result after 3 mappings only needs to be a seq)

As a reader, the bit I'm not getting here from the xmldoc is whether this is a convenience overload to strongly type things, or whether there is a difference in semantics between me using ofAsyncSeq vs ofAsyncArray given I have an array of Asyncs to hand.

(No need to reply in any major depth, just giving feedback on what enters my head as I read this)

Reply:
These all have been there since the start. I might consider moving some out, but it prevents an extra conversion pass. Plus, it is often not clear how to do this thread-safe, hence the extra helpers.

Besides, we mimic `seq` and they have `toList/Array` etc.

Reply:
Note this is inspired/provoked by a combo of two things:

1. spans of ... `list ... List.map |> List.map |> List.toSeq` perpetually coming up in code reviews

2. In Equinox, there's a place where the downstream logic requires a materialized result, but early on, the primary need is to know wheter it is empty or not. Only after some further processing are the actual items required.

There's also an Async angle - the function could be AsyncSeq in nature in that the processing might need to do Async things. But ultimately we need to materialize all the results, so taking an AsyncSeq would not be useful.

Based on this high level requirement, I said "that's a job for a `list`, they have neat comprehensions, are idiomatic and immutable" and convey the correct semantic - we need the full thing, and we'll traverse it in stages.

But, there's also a secondary factor - C# only provides seq comprehensions, and forcing Async rather than Task is bad news.

----

The solution?

Well, the lowest common denominator is seq, that would work naturally for C#, and `list` will naturally upcast to that

But, the real need is a) convey we need all the items (they are not going to be pulled incrementally so no point in implementing it lazily) b) with a quick way to check if the result is empty

So, the `list` + `seq` pair of APIs got changed to a single one that takes an array on the basis that:
- it removes the F# vs C# "slighly different interfaces" confusion/pointless variance (though there still is an Async vs Task divide)
- in F#, an array comprehension vs a `seq {` tends to be less messy
- in C# I'm sure they'll have an Array comprehension eventually, and adding an extension method to to a `Enumerable.ToArray` on the Seq is easy to do
- yes, an array is mutable, but in practice we can tell people "well then don't do that" if they actually mutate items (and most people don't think of arrays as mutable in that way) and/or eventually use `iarray` if something like that becomes universal

----

So, I _think_ that having a similar "not perfect but on balance the best choice" approach should be considered. `Async.Parallel` is an example of this (it takes a `seq` but then calls `Seq.toArray` immediately internally). For me, this would resolve the List vs Array vs ResizeArray thing, as long as the the semantics are clear and/or intuitive (I am guessing what they are here) i.e.:
- for a seq, its lazy
- for all the others they are upcast to seq, so don't go modifying/appending to the resizearray while there's a live iterator on it)